### PR TITLE
split edit distance test cases

### PR DESCRIPTION
**Pull request by marehr**
This changes the edit_distance test case into a test template.

### CODE REVIEW

File: test/unit/alignment/pairwise/edit_distance/global_edit_distance_max_errors_unbanded_test.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance/global_edit_distance_max_errors_unbanded_test.cpp]:
```
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------
```

File: test/unit/alignment/pairwise/edit_distance/global_edit_distance_unbanded_test.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance/global_edit_distance_unbanded_test.cpp]:
```
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------
```

File: test/unit/alignment/pairwise/edit_distance/semi_global_edit_distance_max_errors_unbanded_test.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance/semi_global_edit_distance_max_errors_unbanded_test.cpp]:
```
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------
```

File: test/unit/alignment/pairwise/edit_distance/semi_global_edit_distance_unbanded_test.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
+
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance/semi_global_edit_distance_unbanded_test.cpp]:
```
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------
```

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
+#pragma once
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]:
```
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------
```

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+#include <gtest/gtest.h>
+
+#include <seqan3/alignment/matrix/alignment_score_matrix.hpp>
+#include <seqan3/alignment/matrix/alignment_trace_matrix.hpp>
+#include <seqan3/alignment/pairwise/edit_distance_unbanded.hpp>
+
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]: Why the new line?

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+struct test_traits_type
+{
+    using word_type = word_t;
+    using is_semi_global_type = is_semi_global_t;
+};
+
+template <auto _fixture, typename word_t>
+struct global_fixture : public ::testing::Test
+{
+    auto fixture() -> decltype(alignment_fixture{*_fixture}) const &
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]: Any specific reason for using trailing return type? (I don't mind that much, but AFAIK we are not using it anywhere (outside concepts?))

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    std::vector query = fixture.sequence2;
+
+    auto alignment = edit_distance<typename TypeParam::traits_type>(database, query, align_cfg);
+
+    auto && [gapped_database, gapped_query] = alignment.alignment();
+    EXPECT_EQ(std::string{gapped_database | view::to_char}, fixture.aligned_sequence1);
+    EXPECT_EQ(std::string{gapped_query | view::to_char}, fixture.aligned_sequence2);
+}
+
+REGISTER_TYPED_TEST_CASE_P(edit_distance_unbanded, score, score_matrix, trace_matrix, back_coordinate, front_coordinate, alignment);
```
Comment (eseiler) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]: line length

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp (hunk 6 above)
Comment (marehr) [reply test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]: no clue must be some rebase issue

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp (hunk 7 above)
Comment (marehr) [reply test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp]: This is a left-over of an old idea.

You can write something like this to emphasise that you know the return type
```c++
std::vector result = computation(); // I know it is a vector over some elements
```

But you can't omit the template  when defining a function

```c++
std::vector computation() // does not work
{}

std::vector<auto> computation() // also does not work
{}
```

but with the following you can communicate the return type

```c++
auto computation() -> decltype(std::vector{tmp})
{
   // ...
   return tmp;
}
```

At some point we said we wanted to make it more clear what the return types are, but I think this idea never really gained momentum.

**Comment by marehr** [comment]
@eseiler done; I'll create a new PR for the other missing copy-right headers within `test/unit/alignment/*`

**Comment by h-2** [comment]
How does this relate to other refactoring? I am not sure what we agreed on. Let's discuss this on Thursday with @smehringer and @rrahn.

**Comment by marehr** [comment]
:disappointed: I have a couple of PRs that are based on this :disappointed:

**Comment by h-2** [comment]
reassigning this to @rrahn as agreed on Thursday!

**Pull request merged**